src/array/cpu/spmat_op_impl_coo.cc

```diff
@@ -622,35 +632,128 @@ CSRMatrix UnSortedDenseCOOToCSR(const COOMatrix &coo) {
       coo.col_sorted);
 }
 
-}  // namespace
+// complexity: time O(NNZ), space O(1)
+template <typename IdType>
+CSRMatrix UnSortedSmallCOOToCSR(COOMatrix coo) {
+  const int64_t N = coo.num_rows;
+  const int64_t NNZ = coo.row->shape[0];
+  const IdType *row_data = static_cast<IdType *>(coo.row->data);
+  const IdType *col_data = static_cast<IdType *>(coo.col->data);
+  const IdType *data =
+      COOHasData(coo) ? static_cast<IdType *>(coo.data->data) : nullptr;
+  NDArray ret_indptr = NDArray::Empty({N + 1}, coo.row->dtype, coo.row->ctx);
+  NDArray ret_indices = NDArray::Empty({NNZ}, coo.row->dtype, coo.row->ctx);
+  NDArray ret_data = NDArray::Empty({NNZ}, coo.row->dtype, coo.row->ctx);
+  IdType *Bp = static_cast<IdType *>(ret_indptr->data);
+  IdType *Bi = static_cast<IdType *>(ret_indices->data);
+  IdType *Bx = static_cast<IdType *>(ret_data->data);
+
+  // Count elements in each row
+  std::fill(Bp, Bp + N, 0);
+  for (int64_t i = 0; i < NNZ; ++i) {
+    Bp[row_data[i]]++;
+  }
+
+  // Convert to indexes
+  for (IdType i = 0, cumsum = 0; i < N; ++i) {
+    const IdType temp = Bp[i];
+    Bp[i] = cumsum;
+    cumsum += temp;
+  }
+
+  for (int64_t i = 0; i < NNZ; ++i) {
+    const IdType r = row_data[i];
+    Bi[Bp[r]] = col_data[i];
+    Bx[Bp[r]] = data ? data[i] : i;
+    Bp[r]++;
+  }
+
+  // Restore the indptr
+  for (int64_t i = N; i > 0; --i) {
+    Bp[i] = Bp[i - 1];
+  }
+  Bp[0] = 0;
+
+  return CSRMatrix(
+      coo.num_rows, coo.num_cols, ret_indptr, ret_indices, ret_data,
+      coo.col_sorted);
+}
+
+enum class COOToCSRAlg {
+  sorted = 0,
+  unsortedSmall,
+  unsortedSparse,
+  unsortedDense
+};
 
 /**
 Implementation and Complexity details. N: num_nodes, NNZ: num_edges, P:
 num_threads.
```

Comment:
Mention that: Note that the categorization of "small" or "non-small" and "sparse" or "dense" is determined based on heuristic measures of performance.